src/vcpkg/commands.extract.cpp

```diff
@@ -0,0 +1,166 @@
+        }
+
+        return std::count_if(known_common_prefix.begin(), known_common_prefix.end(), is_slash);
+    }
+
+    std::vector<std::pair<Path, Path>> get_archive_deploy_operations(const ExtractedArchive& archive,
+                                                                     StripSetting strip_setting)
+    {
+        std::vector<std::pair<Path, Path>> result;
+
+        const auto temp_dir = archive.temp_path;
+        const auto base_path = archive.base_path;
+        const auto proximate = archive.proximate_to_temp;
+
+        size_t strip_count = strip_setting.mode == StripMode::Automatic ? get_common_directories_count(proximate)
+                                                                        : static_cast<size_t>(strip_setting.count);
+
+        for (const auto& prox_path : proximate)
+        {
+            const auto& prox_str = prox_path.native();
+            auto first = prox_str.data();
+            auto last = first + prox_str.size();
+
+            // strip leading directories equivalent to the number specified
+            for (size_t i = 0; i < strip_count && first != last; ++i)
+            {
+                first = std::find_if(first, last, is_slash);
+                first = std::find_if_not(first, last, is_slash);
+            }
+
+            if (first == last)
+            {
+                continue;
+            }
+
+            result.emplace_back(temp_dir / Path{prox_path},
+                                Path{base_path} / std::string(first, static_cast<size_t>(last - first)));
+        }
+
+        return result;
+    }
+
+    static void extract_and_strip(
+        const Filesystem& fs, const VcpkgPaths& paths, StripSetting strip_setting, Path archive_path, Path output_dir)
+    {
+        auto temp_dir =
+            extract_archive_to_temp_subdirectory(fs, paths.get_tool_cache(), null_sink, archive_path, output_dir);
+
+        ExtractedArchive archive = {
+            temp_dir, output_dir, fs.get_regular_files_recursive_lexically_proximate(temp_dir, VCPKG_LINE_INFO)};
+
+        auto mapping = get_archive_deploy_operations(archive, strip_setting);
+
+        for (const auto& file : mapping)
+        {
+            const auto& source = file.first;
+            const auto& destination = file.second;
+
+            if (!fs.is_directory(destination.parent_path()))
+            {
+                fs.create_directories(destination.parent_path(), VCPKG_LINE_INFO);
+            }
+
+            fs.rename(source, destination, VCPKG_LINE_INFO);
+        }
+
+        fs.remove_all(temp_dir, VCPKG_LINE_INFO);
+    }
+
+    void command_extract_and_exit(const VcpkgCmdArguments& args, const VcpkgPaths& paths)
+    {
+        auto& fs = paths.get_filesystem();
+        auto parse_args = args.parse_arguments(ExtractCommandStructure);
+        auto archive_path = Path{parse_args.command_arguments[0]};
+        auto destination_path = Path{parse_args.command_arguments[1]};
```

Comment:
Maybe you should also check that both paths aren't the same (and maybe also check that the archive isn't in a sub-path of the path that should be archived)?